### PR TITLE
Add pagination to the version history page and improve api performance

### PR DESCRIPTION
**Pull request by sgalsaleh**
#### What type of PR is this?

type::feature

#### What this PR does / why we need it:
Adds pagination to the version history page and improves the admin console API performance

#### Does this PR introduce a user-facing change?
```release-note
Adds pagination to the version history page and improves the admin console API performance.
```

#### Does this PR require documentation?
https://github.com/replicatedhq/replicated-docs/pull/267

#### Demo
https://www.loom.com/share/0a7dd1af4c3246c69633f8f110b33115

### CODE REVIEW

File: web/src/components/apps/AppVersionHistory.jsx [hunk 1]
```diff
@@ -1051,13 +1058,50 @@ class AppVersionHistory extends Component {
     )
   }
 
+  renderOtherAvailableVersions = () => {
+    // This is kinda hacky. This finds the equivalent downstream version because the midstream
```
Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistory.jsx:1062]: this is a pre-existing comment

File: go.mod [hunk 2]
```diff
@@ -96,6 +96,8 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require golang.org/x/mod v0.5.1 // indirect
```
Comment (divolgin) [inline go.mod:99]: i think go mod tidy will remove this

File: pkg/handlers/app.go [hunk 3]
```diff
@@ -301,19 +283,30 @@ func responseAppFromApp(a *apptypes.App) (*types.ResponseApp, error) {
...
+	TotalCount     int64                                `json:"totalCount"`
 }
 
 func (h *Handler) GetAppVersionHistory(w http.ResponseWriter, r *http.Request) {
 	appSlug := mux.Vars(r)["appSlug"]
 
+	pageSize := 20
+	currentPage := 0
+	if val := r.URL.Query().Get("pageSize"); val != "" {
+		pageSize, _ = strconv.Atoi(val)
```
Comment (divolgin) [inline pkg/handlers/app.go:303]: this overrides default on error with 0

File: pkg/store/kotsstore/version_store.go [hunk 4]
```diff
@@ -787,43 +787,15 @@ func (s *KOTSStore) GetAppVersion(appID string, sequence int64) (*versiontypes.A
...
-func (s *KOTSStore) GetLatestAppVersion(appID string, downloadedOnly bool) (*versiontypes.AppVersion, error) {
-	downstreamVersions, err := s.FindDownstreamVersions(appID, downloadedOnly)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to find app downstream versions")
-	}
-	if len(downstreamVersions.AllVersions) == 0 {
-		return nil, errors.New("no app versions found")
-	}
-	return s.GetAppVersion(appID, downstreamVersions.AllVersions[0].ParentSequence)
+	return versions.AllVersions[0].ParentSequence, nil
```
Comment (divolgin) [inline pkg/store/kotsstore/version_store.go:798]: this should check that there is at least one version.

**Review by divolgin — APPROVED** [review]

**Review by divolgin — DISMISSED** [review]

File: pkg/store/kotsstore/version_store.go (hunk 4 above)
Comment (sgalsaleh) [reply pkg/store/kotsstore/version_store.go:798]: the function that retrieves the versions already does

**Review by divolgin — APPROVED** [review]

**Review by divolgin — APPROVED** [review]

**Pull request merged**